nightfall-optimist/src/services/block-assembler.mjs

```diff
@@ -81,14 +73,53 @@ export async function conditionalMakeBlock(proposer) {
     or we're no-longer the proposer (boo).
    */
   if (proposer.isMe) {
-    const unprocessed = await numberOfUnprocessedTransactions();
-    let numberOfProposableL2Blocks = Math.floor(unprocessed / TRANSACTIONS_PER_BLOCK);
-    // if we want to make a block right now but there aren't enough transactions, this logic
-    // tells us to go anyway
-    if (makeNow && unprocessed > 0 && numberOfProposableL2Blocks === 0)
-      numberOfProposableL2Blocks = 1;
-
-    if (numberOfProposableL2Blocks >= 1) {
+    // Get all the mempool transactions sorted by fee
+    const mempoolTransactions = await getSortedMempoolTransactions();
+
+    // Map each mempool transaction to their byte size
+    const mempoolTransactionSizes = mempoolTransactions.map(tx => {
+      const txSlots =
+        MINIMUM_TRANSACTION_SLOTS +
+        tx.nullifiers.length +
+        Math.ceil(tx.historicRootBlockNumberL2.length / 4) +
+        tx.commitments.length;
+
+      return txSlots * 32;
+    });
+
+    // Calculate the total number of bytes that are in the mempool
+    const totalBytes = mempoolTransactionSizes.reduce((acc, curr) => acc + curr, 0);
+
+    logger.info({
+      msg: 'In the mempool there are the following number of transactions',
+      numberTransactions: mempoolTransactions.length,
+      totalBytes,
+      makeNow,
+    });
+
+    const transactionBatches = [];
+    if (totalBytes > 0) {
+      let bytesSum = 0;
+      for (let i = 0; i < mempoolTransactionSizes.length; ++i) {
+        if (bytesSum + mempoolTransactionSizes[i] > MAX_BLOCK_SIZE) {
+          bytesSum = mempoolTransactionSizes[i];
+          transactionBatches.push(i);
```

Comment:
should be have a `break` after this line at least to terminate the loop early. Otherwise we will continue running the loop despite already reaching the limit for our `transactionBatches`